### ISSUE TITLE
Configure Pylance to not report type hint issues where optionally "None" is returned

### ISSUE THREAD

**Issue by schperplata** [microsoft/pylance-release#371]
Issue Type: <b>Bug</b>

Q: Is there a way to tell Pylance to stop reporting `Optional[]` types? Example:
![image](https://user-images.githubusercontent.com/8736111/93176854-2f2b1700-f732-11ea-9300-07832f29d73d.png)
![image](https://user-images.githubusercontent.com/8736111/93177775-95646980-f733-11ea-8f09-d531a9ea6883.png)

I have functions that accept/return `Optional[str]`, `Optional[bool]`, ... and I would like to be warned only when I pass non-valid type, but not `None`.

Example: 
1. Some path getter, which return path if local folder exists, None otherwise:
```
def getMyPath() -> Optional[str]:
    path = "C:\\asd"
    if os.path.exists(path):
        return path
    else:
        return None
```
2. How this funciton is used.
Note: I am fine if this raise exception. In fact, I want exception on `None` here, otherwise I would `if` it.
```
myPath = getMyPath()
myRelPath = os.path.relpath(myPath, whateva) # <-- error squigles here
```
Now, since `os.path.relpath()` expects `str`, Pylance is warning me that `None` can't be given to `str`. However, I don't want him to do so - specifically for `None` value.

Note: This is a simplified example, in my use case is a bit more complicated, but this should be representative enough.

Extension version: 2020.9.4
VS Code version: Code 1.49.0 (e790b931385d72cf5669fcefc51cdf65990efa5d, 2020-09-10T13:22:08.892Z)
OS version: Windows_NT x64 10.0.18363

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz (8 x 2712)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.92GB (6.43GB free)|
|Process Argv|W:\BlueBox\trunk\fitIDEA3\_fitIDEA3.code-workspace --crash-reporter-id 919e0ed9-46f3-4ad4-9e28-f35755bf8266|
|Screen Reader|no|
|VM|0%|
</details>

**Comment by erictraut** [comment]
Passing None to a function that does not accept None is one of the most common sources of bugs in Python. 

If you are not interested in using Pylance to find bugs in your code, then I recommend disabling type checking by setting `typeCheckingMode` to "off". If you are interested in finding bugs, then I presume you're also interested in fixing those bugs when they're found.

Each time you pass a value that is potentially `None` to a function that is not intended to accept `None`, you are introducing a potential runtime exception.

In your example above, I would recommend adding an assert statement to your code. This is a clue to readers of your code that you intend for `None` to either never occur or to generate a runtime exception. If you include this assert, Pylance will not generate a type mismatch because it will narrow the type of `myPath` to omit the `None` part of the union.

```python
myPath = getMyPath()
assert myPath is not None
myRelPath = os.path.relpath(myPath, whateva)
```

**Comment by schperplata** [comment]
Ok, seems like I would have to write more details.

Let's say I have a class that serves me as a data container, that has a few attributes, which are not initialized by default. Once class instance is created, these attributes are gradually filled and used.
```
class MyData():
    def __init__(self):
        self.myVarOne: str = None
        self.myVarTwo: str = None
```
Pylance complains about these attributes being wrong types:
![image](https://user-images.githubusercontent.com/8736111/93324685-860a1c80-f816-11ea-864d-e504a60423d2.png)

So I do type hints: `Optional[str]`. Pylance is now satisfied. 

Now, I use this class and its attribute values in some function:
```
class MyData():
    def __init__(self):
        self.myVarOne: Optional[str] = None
        self.myVarTwo: str = None


def myFunc(arg: str):
    pass


myData = MyData()
'''
Some code which might or might not set 'myData.myVarOne' value. Might be also in some other function.
'''
myFunc(myData.myVarOne)
```
Now, Pylance is complaining about `None` that can't be casted to `str`:
![image](https://user-images.githubusercontent.com/8736111/93325723-054c2000-f818-11ea-90a3-2dfad857029a.png)


What I want is to decide is whether `None` value has to be explicitly checked or just blindly passed to a function and if anything goes wrong, it will be clearly stated in a default python traceback.  
There are other options:
* add `assert` lines everywhere (a lot of extra lines on many places, with no more info than default traceback)
* chose between Pylance issuing errors in class attributes declaration or function arguments
* disable Pylance type checking error - no way. I still want to be wrned if someone is trying to pass `int` to `str` (but not about `None`).

What would be the proper solution?

**Comment by erictraut** [comment]
Here are a few recommendations, in order of preference.

1. Avoid the pattern of having instance variables that start life as "None" -- i.e. initialize them with a value in the `__init__` method other than `None`. I realize the "initialize with None" pattern is common in Python code, but it's a dangerous pattern that leads to a lot of bugs. If you can avoid it, that's preferable.

2. Use a property pattern where the property's getter asserts that the value does not contain None at the time the value is retrieved. This way, your assert is in one place only.

3. Make your code robust so it properly checks for None at runtime and does something appropriate (other than crashing) in this case. This involves adding `if x is None` statements.

4. Add an `assert` in cases where you are asserting that it is safe to assume that the value does not contain None.

We use all four of these approaches in my team's code base. For new code, we tend to use option 1 and 2 more. For older code that was written before we started using static type checking, we tend to use 3 and 4.

**Comment by schperplata** [comment]
> I realize the "initialize with None" pattern is common in Python code, but it's a dangerous pattern that leads to a lot of bugs. If you can avoid it, that's preferable.
Yeah. This is something our code base is full of, although I don't take it as a such 'buggy' thing - there are a lot more disturbing and easy-to-screw-up things than this. For me: errors where `None` is the reason for crash are much easier to find than if, for example, value of correct type is wrong. And I like `None` in place of a  `str` as a default value, because that way I know that this value has not been set yet. 
Although I appreciate your input and guidelines, this seems to me like it is a much more subjective-type of a design issue, than actual Python code specification. I just hoped that Pylance is somwhat adaptable to my needs.

**Comment by erictraut** [comment]
Embracing static type analysis can require a shift in thinking when writing your code. When properly leveraged, it helps eliminate large classes of runtime bugs.

Disabling selective portions of the type consistency logic is not the right answer here.

**Comment by erictraut** [comment]
Oops, didn't mean to close this.

**Comment by schperplata** [comment]
> Embracing static type analysis can require a shift in thinking when writing your code. When properly leveraged, it helps eliminate large classes of runtime bugs.  
That is the way to use Python as if it is C++.  Which is not. In this specific example, my experiences with data classes with default `None` values are good, and so far I didn't discover that this would be the bottleneck in crashing/debugging. I belive you have different experiences.

> Disabling selective portions of the type consistency logic is not the right answer here.  
I agree. Adapting checker until it is all 'green' is as bad as adapting code design to IDE tools. However, I also agree that static type analysis is basically the only way to make proper sw. 

Feel free to close the issue, thank you for your time.

**Comment by erictraut** [comment]
Let's keep the issue open. I want to give this some more thought. This will also give other Pylance users a chance to comment.

**Comment by oliviercwa** [comment]
One more comment on this one:
Pylance will correctly remove the warning if you assert right before using the variable:
```python
myPath = getMyPath()
assert myPath is not None
myRelPath = os.path.relpath(myPath, whateva)
```

**However**, if you assert in a sub function, the warning will still be there

```python
def assert_not_none(obj: object):
  assert obj is not None

myPath = getMyPath()
assert_not_none(myPath)
myRelPath = os.path.relpath(myPath, whateva)
```

**Comment by jakebailey** [comment]
@oliviercwa That sort of pattern is something PEP 647 is intended to cover (as there's no expectation that the above function body is read to know that it's doing that narrowing): https://www.python.org/dev/peps/pep-0647/

Though we won't have any inference that says "this function is a type guard". That'd be an interesting inference feature.

**Comment by albireox** [comment]
I don't know if this is related, but something that I've seen having an inconsistent behaviour in pylance is how a function argument that is initialised to `None` is treated. In the past (I think), doing this

```python
def f(arg1, some_string: str = None):
    ...
```

would cause a type error because `some_string` was being assigned a `None` value while we were specifying its type as `str`. One would need to change it to `Optional[str]` or to `str | None`. 

In the latest version I'm using (v2021.5.5-pre.1) it seems that the above code is fine for pylance. Is that a bug or an expected feature?

In my opinion this is a desirable feature since I often use `None` as an "undefined" or "not passed" value, and I assume that if `some_string` is not `None` it will always be a string. In that case having to write `Optional[str]` is a bit too verbose and hides the actual type of the argument. But then maybe I'm mis/overusing `None` in these situations.

**Comment by jakebailey** [comment]
That not being errored (if you have type checking enabled) sounds like a bug to me and would be worth a new issue. That form was valid a long time ago when type checking was new, but is not a form that any type checker accepts anymore.

**Comment by albireox** [comment]
Yes, I have type checking set to `basic` so this must be a bug.

**Comment by erictraut** [comment]
I'm able to repro the problem. I'll look into it.

**Comment by erictraut** [comment]
I had forgotten this, but the pyright currently allows this. See [here](https://github.com/microsoft/pyright/blob/9e02e24f9251f2630357dcda03b6e8ff1ce81d2f/packages/pyright-internal/src/analyzer/typeEvaluator.ts#L13319). Two years ago when I added this special case, it was consistent with the behavior prescribed by PEP 484. But the PEP has since been updated to say:

>A past version of this PEP allowed type checkers to assume an optional type when the default value is None ... This is no longer the recommended behavior. Type checkers should move towards requiring the optional type to be made explicit.

Based on this, I think we should change the behavior in pyright. This could result in some new errors in existing code. @jakebailey, any thoughts here?

**Comment by erictraut** [comment]
This is under the control of a flag `strictParameterNoneValue` which is disabled by default for basic type checking mode but enabled by default for strict type checking mode. I propose that we set it to true by default even for basic mode. This would be more consistent with the new guidelines in PEP 484. Users could still disable it if they want.

**Comment by jakebailey** [comment]
I'm not sure how widespread this form is; trying the example with mypy, it doesn't report any errors:

![image](https://user-images.githubusercontent.com/5341706/120029053-6bde2c00-bfaa-11eb-912b-671f9a6437c7.png)

Perhaps we can start by issuing an error that says "don't do this"?

**Comment by albireox** [comment]
My impression, but I may well wrong, is that this has been working a bit inconsistently. I know that at least at some point doing

```python
def f(arg1, some_string: str = None):
    ...
```

did raise a type error. Another thing is that according to the PEP you cited, the behaviour would be to infer that the type is `Optional[str]` but I don't think it does that. If I check the type of `some_string` inside the function, before I override it, I see `str`, not `Optional[str]`.

**Comment by jakebailey** [comment]
In relation to the original issue, though, having this for parameters and not other annotations is perhaps a little incongruent, though, but does match the PEP.

**Comment by erictraut** [comment]
@albireox, the code for "implied optional" types hasn't changed in pyright for the last two years, so if you saw a different behavior perhaps it was because you had enabled "strict" type checking mode.

I confirmed that the type evaluation is consistent.

```python
# pyright: strictParameterNoneValue=false

def f(arg1, some_string: str = None):
    reveal_type(some_string) # str | None
```

```python
# pyright: strictParameterNoneValue=true

def f(arg1, some_string: str = None): # Error
    reveal_type(some_string) # str
```

**Comment by albireox** [comment]
I hardly ever use the strict mode, but I haven't been tracking that consistently so there may have been other issues on my side that confused me.

If `strictParameterNoneValue` is `true` by default in basic type checking, then that matches what I see.

**Comment by stevejpurves** [comment]
> Though we won't have any inference that says "this function is a type guard". That'd be an interesting inference feature.

@jakebailey this is also the case for decorators at the moment, is that true? I attempted to use a decorator to hold my `None` check but pylance still flagged the error

**Comment by jakebailey** [comment]
Unannotated decorators are currently assumed to not affect the signature of the function that they are decorating, yes. See this more complete explanation here: https://github.com/microsoft/pylance-release/issues/1532#issuecomment-875209622

**Comment by IrfanAkram5** [comment]
As a newbie to vs code and python, I find the "None" warnings both useful and irritating!  Would it be possible to add a feature to the effect of "I've checked it and it's OK" to make the warning/errors disappear.  This is more useful than switching off or downgrading a particular check(the squiggly line is just a different colour) as it forces you to check your code.  Sometimes there is nothing you can do because the warning itself is being generated from a library you are using.  E.g. I get the None warning due to the pandas library below.

![image](https://user-images.githubusercontent.com/87219625/125154563-71569680-e152-11eb-94c2-cea54af45b2a.png)

**Comment by erictraut** [comment]
You can disable a warning or error by adding `# type: ignore` to the end of the line where it occurs, but it's better to fix the code such that it handles the `None` case at runtime. You can do so by adding an `if x is not None:` check. Or if you're confident that `None` will never appear in that case, you can add an assert statement (`assert x is not None`).

**Comment by matrs** [comment]
Here, `pyright` is telling me that `"group" is not a known member of "None"`, but i'm ensuring `.group()` isn't used when there is no match (`None`):
```python
imgs_dir_name = (
        re.match(regex, imgs_dir.name).group()
        if re.match(regex, imgs_dir.name)
        else imgs_dir.name
    ) 
```
Could `pyright` understand this at some point or is something that can't be done?

**Comment by matrs** [comment]
you're right, thanks!
